### PR TITLE
Added unit tests for existing components.

### PR DESCRIPTION
**Pull request by SolomiiaLomnytska22**
**Refs:**
#9 (Implement Unit Testing in Project)
**Proposed changes:**
- Added unit testing using Jest.
- Created unit tests for existing components

Testing results:
Screenshot 2023-12-05 at 20 50 47

**Verification:**
To see unit test results:
1. Open project directory.
2. Enter command in terminal "npm run test:unit"

Tested locally. All already existing functionality is supported.

https://github.com/SolomiiaLomnytska22/solomiia-training/assets/136481621/8a10421d-8ec4-49de-a843-2c5417959635

### CODE REVIEW

File: app/package.json [hunk 1]
```diff
@@ -5,38 +5,49 @@
...
   "dependencies": {
+    "@babel/preset-typescript": "^7.23.3",
     "@fortawesome/fontawesome-svg-core": "^6.5.0",
     "@fortawesome/free-brands-svg-icons": "^6.5.0",
     "@fortawesome/free-regular-svg-icons": "^6.5.0",
     "@fortawesome/free-solid-svg-icons": "^6.5.0",
     "@fortawesome/vue-fontawesome": "^2.0.10",
     "axios": "^1.6.1",
     "core-js": "^3.8.3",
+    "flush-promises": "^1.0.2",
```
Comment (marina-grechukha) [inline app/package.json]: where do you need it?

**Comment by kaytee-te** [comment]
A general suggestion about this PR: it's more convenient (especially for large projects) to have test files in the same folder as my Vue components and use the component file name (Button.vue -> Button.spec.ts). This helps to manage component files and test files since they live in the same folder.

File: app/tests/unit/confirm-dialog.spec.ts [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+    wrapper = mount(ConfirmationDialog);
+  });
+
+  afterEach(() => {
+    wrapper.destroy();
+  });
+
+  it('emits confirm event when Confirm button is clicked', async () => {
+    await wrapper.findComponent(Button).trigger('click');
+    await wrapper.vm.$nextTick();
```
Comment (kaytee-te) [inline app/tests/unit/confirm-dialog.spec.ts]: You would either need to `await` for the async `trigger` method or `wrapper.vm.$nextTick()`, no need to have both.

File: app/tests/unit/date-input.spec.ts [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+      propsData: { label, value, id, required, maxDate, minDate }
+    })
+  })
+
+  afterEach(() => {
+    wrapper.destroy()
+  })
+
+  it('renders label and input with correct attributes', async () => {
+    await wrapper.vm.$nextTick()
```
Comment (kaytee-te) [inline app/tests/unit/date-input.spec.ts]: What is the purpose of `vm.$nextTick()` here?

File: app/tests/unit/date-input.spec.ts [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      wrapper.props('maxDate')
+    )
+    expect(wrapper.find('input').attributes('min')).toBe(
+      wrapper.props('minDate')
+    )
+  })
+
+  it('emits input event when input value changes', async () => {
+    const inputField = wrapper.find('input')
+    ;(inputField.element as HTMLInputElement).value = '2023-01-01'
```
Comment (kaytee-te) [inline app/tests/unit/date-input.spec.ts]: Maybe it makes sense to store the `inputField.element as HTMLInputElement).value` in a separate variable. For more code readability.

File: app/package.json (hunk 1 above)
Comment (SolomiiaLomnytska22) [reply app/package.json]: Removed it in a new pull request.

File: app/tests/unit/confirm-dialog.spec.ts (hunk 2 above)
Comment (SolomiiaLomnytska22) [reply app/tests/unit/confirm-dialog.spec.ts]: I've removed nextTicks where it was possible. However, in some cases, tests don't pass if I remove one of await or await nextTick.

File: app/tests/unit/date-input.spec.ts (hunk 3 above)
Comment (SolomiiaLomnytska22) [reply app/tests/unit/date-input.spec.ts]: Removed it in the recent commit.

File: app/tests/unit/user-info-modal.spec.ts [hunk 5]
```diff
@@ -0,0 +1,73 @@
+import { mount, shallowMount } from '@vue/test-utils'
+import UserInfoModal from '@/components/users/UserInfoModal.vue'
+
+describe('UserInfoModal.vue', () => {
+  it('renders when showModal is true', () => {
+    const wrapper = mount(UserInfoModal, {
+      propsData: {
```
Comment (kaytee-te) [inline app/tests/unit/user-info-modal.spec.ts]: There is no need to create a wrapper each time you need to test it with different props values, here is a [setProps](https://v1.test-utils.vuejs.org/api/wrapper/setprops.html) method that could be applied to a wrapper, which was created only once, inside beforeEach, at the beginning.

File: app/tests/unit/confirm-dialog.spec.ts (hunk 2 above)
Comment (kaytee-te) [reply app/tests/unit/confirm-dialog.spec.ts]: Could you point out this case, please?

**Comment by kaytee-te** [comment]
Good work with tests overall, nice job.

File: app/tests/unit/confirm-dialog.spec.ts (hunk 2 above)
Comment (SolomiiaLomnytska22) [reply app/tests/unit/confirm-dialog.spec.ts]: For example, in file user-info-form.spec.ts in test "adds user when form is submitted for "add" action": 
```
await wrapper.find('form').trigger('submit.prevent');
    await wrapper.vm.$nextTick();
```
When i delete one of the awaits, real data differs from expected.

Comment (kaytee-te) [reply app/tests/unit/confirm-dialog.spec.ts]: In this particular case `await wrapper.vm.$nextTick()` makes sense indeed, due to the asynchronous behavior of mockResolvedValueOnce, where we have a Promise as a result. It ensures any promises are settled before the assertion is made.

Comment (SolomiiaLomnytska22) [reply app/tests/unit/confirm-dialog.spec.ts]: Thanks for the info!

**Comment by marina-grechukha** [comment]
> A general suggestion about this PR: it's more convenient (especially for large projects) to have test files in the same folder as my Vue components and use the component file name (Button.vue -> Button.spec.ts). This helps to manage component files and test files since they live in the same folder.

Agree 👍 
Tests should be located in the component's `__tests__` folder. File names should look like `ComponentName.spec.js`.

example:
Screenshot 2023-12-06 at 16 52 33

File: app/tests/unit/user-info-modal.spec.ts (hunk 5 above)
Comment (SolomiiaLomnytska22) [reply app/tests/unit/user-info-modal.spec.ts]: Added setProps instead of creating wrapper in every test in new commit.

**Comment by SolomiiaLomnytska22** [comment]
> A general suggestion about this PR: it's more convenient (especially for large projects) to have test files in the same folder as my Vue components and use the component file name (Button.vue -> Button.spec.ts). This helps to manage component files and test files since they live in the same folder.



> > A general suggestion about this PR: it's more convenient (especially for large projects) to have test files in the same folder as my Vue components and use the component file name (Button.vue -> Button.spec.ts). This helps to manage component files and test files since they live in the same folder.
> 
> Agree 👍 Tests should be located in the component's `__tests__` folder. File names should look like `ComponentName.spec.js`.
> 
> example: Screenshot 2023-12-06 at 16 52 33

Tried to change naming and use folder organisation as in your example in the recent commit, @marina-grechukha, @kaytee-te .

**Review by kaytee-te — COMMENTED** [review]
LGTM

**Review by marina-grechukha — APPROVED** [review]

**Review by kaytee-te — APPROVED** [review]

**Pull request merged**